### PR TITLE
[CPU][BLAS] Fix int overflow in bf16/fp16 gemm staging buffer size

### LINKED ISSUE #191083: CPU bf16/fp16 matmul via sbgemm_/shgemm_ throws on legal shapes: `int c_size = n_ * m_` overflows for outputs >= 2^31 elements
### 🐛 Describe the bug

In `aten/src/ATen/native/CPUBlas.cpp`, the bf16 (`sbgemm_`) and fp16 (`shgemm_`) external-BLAS copy-back paths compute the float staging buffer size in 32-bit `int`:

```cpp
int c_size = n_ * m_;
// C matrix in OpenBLAS sbgemm are of type "float" so we have to convert, copy and copy back.
std::vector<float> float_v(c_size, 0.0f);
```

(`https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/CPUBlas.cpp` — the `BLAS_HAS_SBGEMM` block, ~line 373, and its `BLAS_HAS_SHGEMM` twin, ~line 434.)

`use_blas_gemm` guards each of `m, n, k, lda, ldb, ldc <= INT_MAX` individually, but not the **product** `m*n`. For a legal bf16 matmul whose output has ≥ 2³¹ elements — e.g. 46,341 × 46,341 (each dim tiny relative to INT_MAX; product 2,147,488,281 > 2³¹−1) — `n_ * m_` overflows signed int (UB; wraps negative in practice), and the `std::vector` constructor throws.

Reproduction (CPU build where the BLAS provides `sbgemm_`, i.e. `BLAS_HAS_SBGEMM` is defined — e.g. OpenBLAS with BUILD_BFLOAT16):

```python
import torch
m = n = 46341
k = 64
a = torch.randn(m, k, dtype=torch.bfloat16)
b = torch.randn(k, n, dtype=torch.bfloat16)
c = a @ b   # RuntimeError (std::length_error "vector" from the negative size)
```

Measured on linux-aarch64 (Neoverse V2), OpenBLAS 0.3.33.dev: 46,340² (product just under 2³¹) completes and matches an fp32 reference; 46,341² throws immediately.

### Expected behavior

Either compute `c_size` in `int64_t` (a 2.1G-element float staging buffer is a legal ~8.6 GB allocation, and the surrounding copy loops already index with `int64_t` loop variables), or route these shapes to the internal gemm fallback the way the per-dimension guard already does.

### Versions

PyTorch 2.14 dev (main-lineage, ≈`6a641f6777594fcd2f34ea32f7ee2c0cdaa55776`), linux-aarch64 (Neoverse V2), OpenBLAS 0.3.33.dev LP64 with `sbgemm_` exported. Related to the wider 32-bit-LAPACK/BLAS discussion in #51720 / #190509 (this one is torch-side arithmetic, not LAPACK-side).

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @jianyuh @nikitaved @mruberry @walterddr @Lezcano @snadampal @milpuz01 @nikhil-arm @fadara01 @robert-hardwick @nWEIdia

### PR DESCRIPTION
**Pull request by Diacod-I**
Fixes #191083.

## Problem
In `aten/src/ATen/native/CPUBlas.cpp`, the `sbgemm_` (bf16) and `shgemm_` (fp16) external-BLAS copy-back paths size the fp32 staging buffer with a 32-bit `int`:

```cpp
int c_size = n_ * m_;
std::vector<float> float_v(c_size, 0.0f);
```

`use_blas_gemm` guards each of `m, n, k, lda, ldb, ldc` against `INT_MAX` individually, but not the product `m * n`. For a legal matmul whose output has ≥ 2³¹ elements, e.g. `46341 × 46341` (each dimension small, but product is 2,147,488,281 > 2³¹−1),`n_ * m_` overflows signed `int` and the `std::vector` constructor throws `std::length_error`.

## Fix
Compute the product in `int64_t` by promoting an operand with `static_cast` before the multiply, in both the `sbgemm_` and `shgemm_` blocks. This matches the `int64_t` indexing already used in the surrounding copy loops (`j * m_ + i` with `int64_t` loop variables from `c10::irange`), so the allocation size is now consistent with the indexing.

## Testing note
Reproducing the original throw requires a CPU build whose BLAS exports `sbgemm_`/`shgemm_` (e.g. OpenBLAS built with bf16 support) and an ~8.6 GB fp32 staging allocation for the 2³¹-element case mentioned in the issue, so I wasn't able to reproduce locally. The fix is by inspection: `static_cast<int64_t>(n_) * m_` cannot overflow for the boundary shape the issue documents (46340² works, 46341² throws on the old code). 

I'd appreciate guidance on the preferred way to add a regression test given the allocation size of ~8.6 GB.

### CODE REVIEW

**Comment by Diacod-I** [comment]
This fix is in CPUBlas.cpp, would appreciate if someone could please add the `release notes: cpu` label or the appropriate label on my behalf, since the fix is user facing. Thanks!

**Comment by Skylion007** [comment]
@claude double check this is the right fix and it doesn't get downcast back to int32 implicitly by BLAS

**Review by Skylion007 — APPROVED** [review]

**Comment by Diacod-I** [comment]
@Skylion007
Thank you so much for approving! Seems like the merge failed due to an infra repo fetch failure on a runner-determinator job. Could you re-run `@pytorchbot merge` when you have a moment? Happy to do anything from my end to resolve this. Thank you in advance.
